### PR TITLE
Fix 'type_in_documentation'

### PR DESCRIPTION
**Pull request by texasaggie97**
- [X] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nimi-python/blob/master/CONTRIBUTING.md).
- [X] ~~I've updated [CHANGELOG.md](https://github.com/ni/nimi-python/blob/master/CHANGELOG.md) if applicable.~~
- [X] ~~I've added tests applicable for this pull request~~

### What does this Pull Request accomplish?
* `'type_in_documentation'` would be ignored under certain circumstances (I.e. when using arrays)
    * Mark when we calculate the value instead of it being explicitly set in metadata
    * Use this when figuring out what to put in documentation. If it was explicitly set, this gives it a higher priority than if it was calculated
### What testing has been done?
* Visual inspection of generated files

### CODE REVIEW

File: build/helper/metadata_add_all.py [hunk 1]
```diff
@@ -75,7 +75,10 @@ def _add_python_type(item, config):
 
     # If 'type_in_documentation' isn't in the item, use 'python_type'
     if 'type_in_documentation' not in item:
-        item['type_in_documentation'] = item['python_type']
+        # This is kinda hacky, but we need to know later if the 'type_in_documentation'
```
Comment (marcoskirsch) [inline build/helper/metadata_add_all.py]: "kinda" :)

File: docs/nitclk/class.rst [hunk 2]
```diff
@@ -160,7 +160,7 @@ configure_for_homogeneous_triggers
             
 
 
-        :type sessions: list of (Driver Session or nitclk.SessionReference)
+        :type sessions: (Driver Session or nitclk.SessionReference)
```
Comment (marcoskirsch) [inline docs/nitclk/class.rst]: I think this is now wrong...

File: docs/nitclk/class.rst [hunk 3]
```diff
@@ -460,17 +460,17 @@ exported_sync_pulse_output_terminal
...
-            | Permissions    | read-write |
-            +----------------+------------+
-            | Channel Based  | No         |
-            +----------------+------------+
-            | Resettable     | No         |
-            +----------------+------------+
+            +----------------+-------------+
+            | Characteristic | Value       |
+            +================+=============+
+            | Datatype       | default_str |
```
Comment (marcoskirsch) [inline docs/nitclk/class.rst]: this is wrong

File: generated/nidcpower/nidcpower/session.py [hunk 4]
```diff
@@ -118,7 +118,7 @@ class _SessionBase(object):
     _is_frozen = False
 
     _active_advanced_sequence = _attributes.AttributeViString(1150074)
-    '''Type: str
+    '''Type: default_str
```
Comment (marcoskirsch) [inline generated/nidcpower/nidcpower/session.py]: These are all wrong

File: docs/nitclk/class.rst (hunk 2 above)
Comment (texasaggie97) [reply docs/nitclk/class.rst]: Yes, missed adding this in the nitclk metadata now that "list of " no longer added.

File: docs/nitclk/class.rst (hunk 3 above)
Comment (texasaggie97) [reply docs/nitclk/class.rst]: Lost fix when moving to new PR. Fixed. Again.

File: generated/nidcpower/nidcpower/session.py (hunk 4 above)
Comment (texasaggie97) [reply generated/nidcpower/nidcpower/session.py]: Lost fix when moving to new PR. Fixed. Again.

File: build/helper/documentation_helper.py [hunk 5]
```diff
@@ -485,14 +485,16 @@ def format_type_for_rst_documentation(param, numpy, config):
...
-        p_type = 'numpy.array(dtype=numpy.{0})'.format(get_numpy_type_for_api_type(param['type'], config))
-    elif param['use_list'] is True:
-        p_type = 'list of ' + p_type
-    elif param['use_array'] is True:
-        p_type = 'array.array("{0}")'.format(get_array_type_for_api_type(param['type']))
+    # If type_in_documentation was set in metadata, we use it as is
+    if param['type_in_documentation_was_calculated'] or numpy:
+        if param['is_string'] is True:
+            p_type = 'str'
+        elif param['is_buffer'] is True and numpy is True:
```
Comment (marcoskirsch) [inline build/helper/documentation_helper.py]: Is it possible for numpy to be True but `param['is_buffer']` is False? It seems to me like that would be a metadata mistake. I think we should capture that in code by removing `is_buffer` from the condition and making it an assert.

File: build/helper/documentation_helper.py [hunk 6]
```diff
@@ -561,14 +563,16 @@ def _format_type_for_docstring(param, numpy, config):
...
+            p_type = 'numpy.array(dtype=numpy.{0})'.format(get_numpy_type_for_api_type(param['type'], config))
```
Comment (marcoskirsch) [inline build/helper/documentation_helper.py]: same as above.

File: docs/nitclk/class.rst [hunk 7]
```diff
@@ -183,7 +183,7 @@ finish_sync_pulse_sender_synchronize
             
 
 
-        :type sessions: list of (nimi-python Session class or nitclk.SessionReference)
+        :type sessions: (nimi-python Session class or nitclk.SessionReference)
```
Comment (marcoskirsch) [inline docs/nitclk/class.rst]: this is still wrong, should be "list of..."

File: generated/nitclk/nitclk/session.py [hunk 8]
```diff
@@ -614,7 +614,7 @@ def finish_sync_pulse_sender_synchronize(self, sessions, min_time=datetime.timed
         Finishes synchronizing the Sync Pulse Sender.
 
         Args:
-            sessions (list of (nimi-python Session class or nitclk.SessionReference)): sessions is an array of sessions that are being synchronized.
+            sessions ((nimi-python Session class or nitclk.SessionReference)): sessions is an array of sessions that are being synchronized.
```
Comment (marcoskirsch) [inline generated/nitclk/nitclk/session.py]: wrong

File: generated/nitclk/nitclk/session.py [hunk 9]
```diff
@@ -939,7 +939,7 @@ def finish_sync_pulse_sender_synchronize(sessions, min_time):
     Finishes synchronizing the Sync Pulse Sender.
 
     Args:
-        sessions (list of (nimi-python Session class or nitclk.SessionReference)): sessions is an array of sessions that are being synchronized.
+        sessions ((nimi-python Session class or nitclk.SessionReference)): sessions is an array of sessions that are being synchronized.
```
Comment (marcoskirsch) [inline generated/nitclk/nitclk/session.py]: wrong

**Comment by marcoskirsch** [comment]
Note that codecov didn't run.

File: docs/nitclk/class.rst (hunk 7 above)
Comment (texasaggie97) [reply docs/nitclk/class.rst]: Missed in metadata. Fixed

File: generated/nitclk/nitclk/session.py (hunk 8 above)
Comment (texasaggie97) [reply generated/nitclk/nitclk/session.py]: See above

File: generated/nitclk/nitclk/session.py (hunk 9 above)
Comment (texasaggie97) [reply generated/nitclk/nitclk/session.py]: See above

File: build/helper/documentation_helper.py [hunk 10]
```diff
@@ -485,14 +485,17 @@ def format_type_for_rst_documentation(param, numpy, config):
...
-    if param['is_string'] is True:
-        p_type = 'str'
-    elif param['is_buffer'] is True and numpy is True:
-        p_type = 'numpy.array(dtype=numpy.{0})'.format(get_numpy_type_for_api_type(param['type'], config))
-    elif param['use_list'] is True:
-        p_type = 'list of ' + p_type
-    elif param['use_array'] is True:
-        p_type = 'array.array("{0}")'.format(get_array_type_for_api_type(param['type']))
+    # If 'type_in_documentation' does not start with 'default_', we know that it was
+    # explicitly set in metadata and we do not over write it.
```
[the remark this answers is not in the document]
Comment (texasaggie97) [reply build/helper/documentation_helper.py]: This has already changed. I think you commented on an old version of the file. At least I do not see it anymore.

**Pull request merged**